torch/utils/data/dataset.py

```diff
@@ -260,6 +265,125 @@ def __len__(self):
+    Once a chunk index has been selected, it is passed to ``ChunkDataReader`` that will load
+    data into the ``ChunkDataset`` internal cache. Only then ``ChunkDataset`` returns batches.
+
+    Arguments:
+        chunk_sampler (DistributedSampler or DistributedChunkSampler): Draw indices for chunks.
+            Typically used to split data amongst dataloader workers (first level of sampling)
+        chunk_reader (ChunkDataReader): Specialized reader for a given input type
+        shuffle_cache (bool): Setting `True` trigger shuffling of the internal chunk cache
+            (second level of sampling) (default: `True`)
+    """
+
+    def __init__(self, chunk_sampler, chunk_reader, shuffle_cache=True):
+        super(ChunkDataset, self).__init__()
+        assert isinstance(chunk_sampler, DistributedSampler) or isinstance(
+            chunk_sampler, DistributedChunkSampler), 'sampler must be a `DistributedSampler` or `DistributedChunkSampler`'
+        assert callable(chunk_reader), 'chunk_reader must be `callable()` and return a container with data'
+        assert isinstance(shuffle_cache, bool), 'shuffle_cache must be a `bool`'
+
+        self.chunk_sampler = chunk_sampler
+        self.chunk_reader = chunk_reader
+        self.shuffle_cache = shuffle_cache
+
+        # Internal state
+        self._chunk_sampler_iter = iter(self.chunk_sampler)
+        self._cache = []
+        self._batch_size = 1
+        self._min_cache = 1000
+
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        raise NotImplementedError
+
+    def __next__(self):
+        r"""Returns a batch or raises exception when exhausted"""
+
+        cache_size = 0
+        if len(self._cache) > 0:
+            cache_size = len(self._cache)
+
+        if (cache_size < self._min_cache):
+            new_chunk = None
+            # Try reading a new chunk
+            # Ignore EOF here because there may be internal cache entries to be returned yet
+            try:
+                new_chunk = self.chunk_reader(next(self._chunk_sampler_iter))
+                assert isinstance(new_chunk, list), "Chunk data reader must return a `list(samples)`"
+            except StopIteration:
+                pass
+
+            # Updates internal cache
+            if new_chunk and cache_size > 0:
+                self._cache = self._cache + new_chunk
+            elif new_chunk:
+                self._cache = new_chunk
+            elif not new_chunk and cache_size == 0:
+                raise StopIteration
+
+            # Apply second level of shuffling
+            if self.shuffle_cache:
+                random.shuffle(self._cache)
+
+        # Get a batch and update internal cache cache
+        if len(self._cache) > 0:
+            batch = self._cache[:self._batch_size]
+            self._cache = self._cache[self._batch_size:]
+
+        # `self._batch_size` is hard-coded to 1 to leverage DataLoader's auto collation
```

Comment:
So right now, the output is always going through Dataloader's auto collation code path?

Reply:
yes. Jaliya questioned the same and I changed the comment to make it less confusing. Maybe we should just remove this internal constant and if one day we want ChunkDataset to return a batch, we re-add batch_size as an official input.

@jaliyae @xzhu1900  what do you think?

Reply:
I agree, let's remove this constant.

Reply:
second to this.

Reply:
fixed